### DISCUSSION TITLE
How to handle client interrupts?

### DISCUSSION THREAD

**Discussion by guskovd** [hyperium/tonic#702] [Q&A]
Let's say there is a simple service:

```proto
rpc Some(SomeRequest) returns (SomeResponse)
```

what is converted to

```rust
    async fn some(
        &self,
        request: Request<SomeRequest>,
    ) -> Result<Response<SomeResponse>, Status> {
    ...
    }
```

Service works, everything is fine

But what if I need to handle client interrupts? I tried to simulate this situation and I have a feeling that in this case the `some` method just exits immediately

**Comment by LucioFranco** [comment]
When you say interrupts you mean when the client disconnects? Can you explain your use case a bit more because of the way hyper works its hard to detect when you have on write errors.

**Reply by Evan30s** [reply]
> When you say interrupts you mean when the client disconnects? Can you explain your use case a bit more because of the way hyper works its hard to detect when you have on write errors.

I may found the same problem, I may do many things in `some` function, and I want finish `some` function even the client abort(the user can kill the client thread). However, I fould that the server will not keep runing the `some` function if I kill the client, which may cause some bugs.
For example:
```rust
async fn some(
        &self,
        request: Request<SomeRequest>,
) -> Result<Response<SomeResponse>, Status> {
        let mut i = 0;
        while i < 30 {
            i += 1;
            println!("wait {} seconds", i);
            sleep(Duration::from_secs(1)).await;
        }

        Ok(Response::new(SomeResponse{}))
}
```
If you kill the client after 3s, you will only see below.
- wait 1 seconds
- wait 2seconds
- wait 3 seconds
But you would not see below.
- wait 4 seconds
- wait 5 seconds

**Reply by davidpdrsn** [reply]
If you want things to run outside of the request handler you can use `tokio::spawn`

**Reply by Evan30s** [reply]
> If you want things to run outside of the request handler you can use `tokio::spawn`

It can work, but it may cause the code a little strange. Also, something I also want to call some functions of self, so I have to wrap the self with Arc, which make the code wierd. So I hope it could be one way that I can control the request handler would not be aborted even the client abort.

**Reply by davidpdrsn** [reply]
Having to clone self into spawn seems fine to me. I'm not sure we need something dedicated for this.
